### PR TITLE
ref(native): Implement a UI for debug sources

### PR DESCRIPTION
**Pull request by jan-auer**
This PR implements a proper UI for configuring debug sources and also adds server-side support for HTTP and GCS sources. A couple of things to note:

 - There is no more way to edit the JSON. In the future, it might make sense to have a UI where one can copy-paste JSON directly. 
- Source identifiers are generated randomly and immutable. They are not changed when the source changes.
 - Undo is broken, as with all fields that use `setValue` / `getValue`.

Screenshot 2019-07-04 at 15 18 39

### CODE REVIEW

File: src/sentry/static/sentry/app/components/modals/debugFileSourceModal.jsx [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    placeholder: 'gcs-bucket-name',
+    help: t('Name of the GCS bucket. Read permissions are required to download symbols.'),
+  },
+  clientEmail: {
+    name: 'client_email',
+    type: 'email',
+    required: true,
+    label: t('Client Email'),
+    placeholder: 'user@project.iam.gserviceaccount.com',
+    help: t('E-Mail address of the GCS service account.'),
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/modals/debugFileSourceModal.jsx]: You mix usage of `E-Mail` and `Email` can we use one form?

File: src/sentry/static/sentry/app/components/modals/debugFileSourceModal.jsx [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    required: true,
+    multiline: true,
+    autosize: true,
+    maxRows: 5,
+    label: t('Private Key'),
+    placeholder: '-----BEGIN PRIVATE KEY-----\n[PRIVATE-KEY]\n-----END PRIVATE KEY-----',
+    help: tct('The service account key. Credentials can be managed on the [link].', {
+      link: (
+        <ExternalLink href="https://console.cloud.google.com/project/_/iam-admin">
+          IAM & Admin Page
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/modals/debugFileSourceModal.jsx]:
```suggestion
          IAM &amp; Admin Page
```

File: src/sentry/static/sentry/app/data/forms/projectDebugFiles.jsx [hunk 3]
```diff
@@ -1,8 +1,45 @@
...
+
+export function getSourceName(type) {
+  return sourceNames[type] || t('Unknown');
+}
+
+function flattenKeys(obj) {
+  const result = {};
+  _.forEach(obj, (value, key) => {
+    if (_.isObject(value)) {
+      _.forEach(value, (innerValue, innerKey) => {
```
Comment (markstory) [inline src/sentry/static/sentry/app/data/forms/projectDebugFiles.jsx]: Do you ever need to recursively flatten keys beyond 2 levels?

File: src/sentry/static/sentry/app/data/forms/projectDebugFiles.jsx [hunk 4]
```diff
@@ -18,30 +55,74 @@ export const fields = {
...
+        sourceType: item.type,
+        onSave: updateItem,
+      });
+    },
+
+    removeConfirm: {
+      title: t('Remove Repository?'),
+      confirmText: t('Remove Repository'),
+      message: (
+        <>
```
Comment (markstory) [inline src/sentry/static/sentry/app/data/forms/projectDebugFiles.jsx]: Are we using this syntax? In most of the application we use `<React.Fragment>` as `<>` is un-googlable to folks who don't know what it is.

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+const ItemLabel = styled('span')`
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  padding: 8px 12px;
+  white-space: nowrap;
+`;
+
+const ItemIcon = styled('span')`
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: You might want to use a styled(Button) here. You'll get accessibility and pointer events for free, and the `borderless size="zero"` mode might get you close to what you have here.

File: src/sentry/static/sentry/app/data/forms/projectDebugFiles.jsx (hunk 3 above)
Comment (jan-auer) [reply src/sentry/static/sentry/app/data/forms/projectDebugFiles.jsx]: No, we only have one level of nesting in these objects. 

To be fair, this is just a hack because our form models cannot deal with nested objects. Ideally, I would like to get rid of this completely.

File: src/sentry/static/sentry/app/data/forms/projectDebugFiles.jsx (hunk 4 above)
Comment (jan-auer) [reply src/sentry/static/sentry/app/data/forms/projectDebugFiles.jsx]: @EvanPurkhiser or @billyvg, can you please inform on this?

My take on this is that developers touching this should know what they are doing and `React.Fragment` is just lengthy, but I'll let you decide on this.

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx (hunk 5 above)
Comment (jan-auer) [reply src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: I played around with this and wasn't able to make it work (without assuming too much about the internal implementation of Button). Suggestions welcome!

Comment (markstory) [reply src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: I tried to get our current button component to fit and it was pretty clunky to do with styled components. I could add a new `micro` button size though that gets you the styling you're looking for if that would help.

Comment (jan-auer) [reply src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: Sounds good, thanks! 👍

File: src/sentry/static/sentry/app/constants/index.jsx [hunk 6]
```diff
@@ -157,3 +157,41 @@ export const SEARCH_TYPES = {
...
+  ['cn-northwest-1', 'China (Ningxia)'],
+  ['eu-central-1', 'EU (Frankfurt)'],
+  ['eu-west-1', 'EU (Ireland)'],
+  ['eu-west-2', 'EU (London)'],
+  ['eu-west-3', 'EU (Paris)'],
+  ['eu-north-1', 'EU (Stockholm)'],
+  ['sa-east-1', 'South America (São Paulo)'],
+  ['us-gov-east-1', 'AWS GovCloud (US-East)'],
+  ['us-gov-west-1', 'AWS GovCloud (US)'],
+];
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/constants/index.jsx]: Would it make more sense for this to live in `app/data/debugSources` ?

File: src/sentry/static/sentry/app/data/forms/projectDebugFiles.jsx [hunk 7]
```diff
@@ -18,30 +55,74 @@ export const fields = {
...
+    },
+
+    getValue: sources => JSON.stringify(sources.map(unflattenKeys)),
+    setValue: raw => (JSON.parse(raw || null) || []).map(flattenKeys),
+
+    renderItem(item) {
+      if (item.name) {
+        return item.name;
+      } else {
+        return <em>{t('<Unnamed Repository>')}</em>;
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/data/forms/projectDebugFiles.jsx]: total 🚲 🏠 but could just write `return item.name || <em>{t('<Unnamed Repository>')}</em>;`

File: src/sentry/static/sentry/app/data/forms/projectDebugFiles.jsx (hunk 4 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/data/forms/projectDebugFiles.jsx]: 👎 on this syntax for the moment. Until we have a eslint rule to be consistent, lets stick with `React.Fragment`.

File: src/sentry/static/sentry/app/views/settings/components/forms/fieldFromConfig.jsx [hunk 8]
```diff
@@ -69,6 +73,8 @@ export default class FieldFromConfig extends React.Component {
     };
 
     switch (field.type) {
+      case 'separator':
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/forms/fieldFromConfig.jsx]: omg we've come full circle and are now implementing windows forms.

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+import InputField from 'app/views/settings/components/forms/inputField';
+import InlineSvg from 'app/components/inlineSvg';
+import Confirm from 'app/components/confirm';
+
+const ItemList = styled('div')`
+  display: flex;
+  flex-wrap: wrap;
+`;
+
+const Item = styled('span')`
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: Why span and not a div?

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+import styled from 'react-emotion';
+
+import {t} from 'app/locale';
+import DropdownAutoComplete from 'app/components/dropdownAutoComplete';
+import DropdownButton from 'app/components/dropdownButton';
+import InputField from 'app/views/settings/components/forms/inputField';
+import InlineSvg from 'app/components/inlineSvg';
+import Confirm from 'app/components/confirm';
+
+const ItemList = styled('div')`
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: Typically we stick all our styled components at the bottom of the file, after the main component definition.

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+   * removed immediately.
+   */
+  removeConfirm: PropTypes.object,
+};
+
+function getDefinedProps(propTypes, props) {
+  return _.pickBy(props, (_prop, key) => key in propTypes);
+}
+
+class RichList extends React.PureComponent {
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: 👍 for the field name. We probably should stick `Field` as a suffix to be consistent (though I know the one I built didn't do this)

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+     * The list of items to render.
+     */
+    value: PropTypes.array.isRequired,
+  };
+
+  static defaultProps = {
+    addButtonText: t('Add Item'),
+    renderItem: item => item,
+    onAddItem: (item, addItem) => addItem(item),
+    onEditItem: () => {},
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: maybe passing null to this could disable the edit gear?

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+      <Confirm
+        priority="danger"
+        confirmText={t('Remove')}
+        {...this.props.removeConfirm}
+        onConfirm={() => this.onRemoveItem(item, index)}
+      >
+        {removeIcon}
+      </Confirm>
+    ) : (
+      React.cloneElement(removeIcon, {
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: Maybe instead of cloning just make the `removeIcon` a function that takes some extra props? Slightly less confusing I think.

Could probably pull it out of this class body entirely as well.

File: src/sentry/static/sentry/app/views/settings/components/forms/fieldSeparator.jsx [hunk 14]
```diff
@@ -0,0 +1,5 @@
+import React from 'react';
+
+export default function FieldSeparator(props) {
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/forms/fieldSeparator.jsx]: Would be nice to see this in a separate PR.

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+
+import {t} from 'app/locale';
+import DropdownAutoComplete from 'app/components/dropdownAutoComplete';
+import DropdownButton from 'app/components/dropdownButton';
+import InputField from 'app/views/settings/components/forms/inputField';
+import InlineSvg from 'app/components/inlineSvg';
+import Confirm from 'app/components/confirm';
+
+const ItemList = styled('div')`
+  display: flex;
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: I would actually recommend using display grid here, with auto flow and grid-gap, you'll reduce a lot of the spacing styling you need in `Item` and I think can also probably completely get rid of `ItemLabel`.

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+      </Confirm>
+    ) : (
+      React.cloneElement(removeIcon, {
+        onClick: () => this.onRemoveItem(item, index),
+      })
+    );
+
+    return (
+      <Item size="small" key={index}>
+        <ItemLabel>
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: I think we could collapse a lot of these DOM elements down.

Why do we need `Item` and `ItemLabel`, can we not make `Item` style things like `ItemLabel` does? Why do we need `ItemIcon`? Why can't `ItemIcon` just extend InlineSvg and apply some styling?

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+            <InlineSvg src="icon-settings" size="12px" />
+          </ItemIcon>
+          {removeConfirm}
+        </ItemLabel>
+      </Item>
+    );
+  };
+
+  renderDropdown = () => {
+    return (
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: 🚲 🏠 you can get rid of the return and brace above and reduce your indention one level.

Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: Do we know why there's a weird gap in space between the dropdown in the screenshot?

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+  };
+
+  renderRichList = fieldProps => {
+    const richListProps = getDefinedProps(RichListProps, this.props);
+    const {value, ...props} = fieldProps;
+
+    // We must not render this field until `setValue` has been applied by the
+    // model, which is done after the field is mounted for the first time. To
+    // check this, we cannot use Array.isArray because the value passed in by
+    // the model might actually be an ObservableArray.
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: 🤔 why?

**Review by evanpurkhiser — CHANGES_REQUESTED** [review]

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx (hunk 13 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: Or just make it a component even.

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx (hunk 18 above)
Comment (mitsuhiko) [reply src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: The answer to this question would only unsettle you.

Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: 😬

File: src/sentry/static/sentry/app/components/modals/debugFileSourceModal.jsx [hunk 19]
```diff
@@ -0,0 +1,273 @@
...
+import SentryTypes from 'app/sentryTypes';
+import {objectToArray} from 'app/utils';
+
+import {getSourceName} from 'app/data/forms/projectDebugFiles';
+import Form from 'app/views/settings/components/forms/form';
+import FieldFromConfig from 'app/views/settings/components/forms/fieldFromConfig';
+import ExternalLink from '../links/externalLink';
+
+function objectToChoices(obj) {
+  return objectToArray(obj).map(([key, value]) => [key, t(value)]);
```
[the remark this answers is not in the document]
Comment (jan-auer) [reply src/sentry/static/sentry/app/components/modals/debugFileSourceModal.jsx]: Yes it does, and we use both. Is there a preference for `Object.entries`?

File: src/sentry/static/sentry/app/components/modals/debugFileSourceModal.jsx [hunk 20]
```diff
@@ -0,0 +1,273 @@
...
+
+const commonFields = {
+  id: {
+    name: 'id',
+    type: 'hidden',
+    required: true,
+    defaultValue: () =>
+      Math.random()
+        .toString(36)
+        .substring(2),
```
[the remark this answers is not in the document]
Comment (jan-auer) [reply src/sentry/static/sentry/app/components/modals/debugFileSourceModal.jsx]: This creates a random string to initialize the ID with. It needs to be random within the project.

File: src/sentry/static/sentry/app/views/settings/components/forms/fieldFromConfig.jsx (hunk 8 above)
Comment (jan-auer) [reply src/sentry/static/sentry/app/views/settings/components/forms/fieldFromConfig.jsx]: Amazing times we live in.

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx (hunk 9 above)
Comment (jan-auer) [reply src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: Thinking about it, this should actually be a `<li>`.

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx (hunk 11 above)
Comment (jan-auer) [reply src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: This is not the field. This is the list within the field. The field is actually called `RichListField`.

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx (hunk 13 above)
Comment (jan-auer) [reply src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: What's the problem with cloning here? This is exactly what `cloneElement` is for. 
Going with the function then.

File: src/sentry/static/sentry/app/views/settings/components/forms/fieldSeparator.jsx (hunk 14 above)
Comment (jan-auer) [reply src/sentry/static/sentry/app/views/settings/components/forms/fieldSeparator.jsx]: Noted. Next time (and especially when not dealing with the settings UI). I need to get this out and don't want to start the rebase game anymore.

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx (hunk 15 above)
Comment (jan-auer) [reply src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: Wait, `ItemLabel` does not interact with `ItemList`. I removed some redundant rules now, they were leftovers from other experiments.

FWIW: The `Item` styles, including the label, are mostly copied from our `Button` implementation, which I could massage into rendering like this. I'd rather keep it close to how Button is structured.

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx (hunk 16 above)
Comment (jan-auer) [reply src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: See previous comment. That was taken from `Button`, which is not flexible enough to be used in this scenario.

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx (hunk 17 above)
Comment (jan-auer) [reply src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: Would that be the preferred style? Imho this is inconsistent with other lexically bound methods.

Comment (jan-auer) [reply src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: Due to the flex container, which is resizing the dropdown's div container. Might go away with switching to grid. Let me play around with this.

Comment (jan-auer) [reply src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: Fixed in the next commit.

File: src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx (hunk 18 above)
Comment (jan-auer) [reply src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx]: The idiocracy of how fields interact with the model, instead of a simple top-down data flow:
 1. The form component renders with initial data. Fields are not rendered at this point.
 2. The model initializes its fields. Fields are still not rendered, which means field props are not available.
 3. Fields render. After they mount, they notify the model about their props.
 4. The model adjusts to the props, and calls `setValue`. 

This means that the initial render of the fields is with plain wrong data. This happens in ProjectGeneralSettings, for instance. Oh, and undo is broken for all fields that use `getValue` / `setValue` for a similar reason.

File: src/sentry/static/sentry/app/components/modals/debugFileSourceModal.jsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/components/modals/debugFileSourceModal.jsx]: No preference if we're already inconsistent

**Review by evanpurkhiser — APPROVED** [review]
Still have reservations about the deep nesting of DOM elements. IMO button has that issue as well tbh, so just because button does it doesn't mean it should be done here.

Do you know why there's this gap here?
image




Fantastic work on this

**Comment by jan-auer** [comment]
@EvanPurkhiser thanks! Will simplify the styles and merge afterwards. The gap is gone already, didn’t update the screenshot anymore. 
Thanks @markstory for adding that button size.
